src/lib/Microsoft.Health.Fhir.Ingest/Service/NormalizationEventConsumerService.cs

```diff
@@ -0,0 +1,218 @@
+            NormalizationExceptionTelemetryProcessor exceptionTelemetryProcessor)
+        {
+            _converter = EnsureArg.IsNotNull(converter, nameof(converter));
+            _templateDefinition = EnsureArg.IsNotNullOrWhiteSpace(templateDefinition, nameof(templateDefinition));
+            _templateManager = EnsureArg.IsNotNull(templateManager, nameof(templateManager));
+            _collector = EnsureArg.IsNotNull(collector, nameof(collector));
+            _logger = EnsureArg.IsNotNull(logger, nameof(logger));
+            _retryPolicy = CreateRetryPolicy(logger);
+            _collectionTemplateFactory = EnsureArg.IsNotNull(collectionTemplateFactory, nameof(collectionTemplateFactory));
+            _exceptionTelemetryProcessor = EnsureArg.IsNotNull(exceptionTelemetryProcessor, nameof(exceptionTelemetryProcessor));
+
+            EventMetrics.SetConnectorOperation(ConnectorOperation.Normalization);
+        }
+
+        private (IContentTemplate template, DateTimeOffset timestamp) NormalizationTemplate { get; set; }
+
+        public async Task ConsumeAsync(IEnumerable<IEventMessage> events)
+        {
+            EnsureArg.IsNotNull(events);
+
+            await _retryPolicy.ExecuteAsync(async () => await ConsumeAsyncImpl(events));
+        }
+
+        private async Task<IContentTemplate> GetNormalizationTemplate()
+        {
+            await semaphore.WaitAsync();
+            try
+            {
+                if (NormalizationTemplate.template == null)
+                {
+                    _logger.LogTrace("Initializing normalization template from blob.");
+                    DateTimeOffset timestamp = DateTimeOffset.UtcNow;
+                    var content = await _templateManager.GetTemplateContentIfChangedSince(_templateDefinition);
+                    var templateContext = _collectionTemplateFactory.Create(content);
+                    templateContext.EnsureValid();
+                    NormalizationTemplate = (templateContext.Template, timestamp);
+                }
+                else
+                {
+                    DateTimeOffset updatedTimestamp = DateTimeOffset.UtcNow;
+                    var content = await _templateManager.GetTemplateContentIfChangedSince(_templateDefinition, NormalizationTemplate.timestamp);
+
+                    if (content != null)
+                    {
+                        _logger.LogTrace("New normalization template content detected, updating template.");
+                        var templateContext = _collectionTemplateFactory.Create(content);
+                        templateContext.EnsureValid();
+                        NormalizationTemplate = (templateContext.Template, updatedTimestamp);
+                    }
+                }
+
+                return NormalizationTemplate.template;
+            }
+            finally
+            {
+                semaphore.Release();
+            }
+        }
```

Comment:
nit: Can this method (GetNormalizationTemplate()) and the semaphore be moved to the TemplateManager? Just curious if we can encapsulate all the logic that determines which template to use in a single class that we can update later if we want to.

Reply:
The template manager is also used for the FHIR mapping templates.  I didn't want to impact that code significantly with this PR.  But you are correct, this can be and should be handled by it's own class/responsibility.  I am planning on addressing this in a follow up PR (but it won't be part of next week's release).